samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java

```diff
@@ -524,10 +468,125 @@ public void onStreamProcessorLaunchFailure(SamzaResource resource, Throwable t)
+                  "after a retry delay of {} ms. " +
+                  "This is greater than max retry count of {} and max retry window of {} ms, " +
+                  "so shutting down the application master and marking the job as failed.",
+              processorId, containerId, maxRetryCount, currentRetryWindowMs, retryDelay.toMillis(), maxRetryCount, maxRetryWindowMs);
+
+          // We have too many failures, and we're within the window
+          // boundary, so reset shut down the app master.
+          tooManyFailedContainers = true;
+          state.status = SamzaApplicationState.SamzaAppStatus.FAILED;
+        } else {
+          LOG.info("Resetting failure count for Processor ID: {} back to 1, since last failure " +
+              "(for Container ID: {}) was outside the bounds of the retry window.", processorId, containerId);
+
+          // Reset counter back to 1, since the last failure for this
+          // container happened outside the window boundary.
+          processorFailures.put(processorId, new ProcessorFailure(1, Instant.now()));
+        }
+      } else {
+        LOG.info("Current failure count for Processor ID: {} is {}.", processorId, currentFailCount);
+        processorFailures.put(processorId, new ProcessorFailure(currentFailCount, Instant.now()));
+      }
+    }
+
+    if (!tooManyFailedContainers) {
+      Duration retryDelay = getHostRetryDelay(lastSeenOn, currentFailCount);
+      LOG.info("Retrying request for preferred host: {} for Processor ID: {} (current Container ID: {}) with a delay of {} ms.",
+          lastSeenOn, processorId, resourceStatus.getContainerId(), retryDelay.toMillis());
+      handleContainerStop(processorId, resourceStatus.getContainerId(), lastSeenOn, exitStatus, retryDelay);
+    }
+  }
+
+  /**
+   * Calculates the container request retry delay based on the host name.
+   * @param host host name
+   * @param failCount current number of times the container on the host failed
+   * @return the duration of the exponential backoff delay calculated from the failCount. The max delay is obtained from
+   *   {@link ClusterManagerConfig#getContainerRetryMaxDelayMs()}. If the host name is equal to
+   *   {@link ResourceRequestState#ANY_HOST}, then always return {@link Duration#ZERO}.
+   */
+  @VisibleForTesting
+  Duration getHostRetryDelay(String host, int failCount) {
+    // Only add a retry delay when host is a preferred host or if the request failed more than once.
+    if (failCount < 2 || StringUtils.isBlank(host) || host.equals(ResourceRequestState.ANY_HOST)) {
+      return Duration.ZERO;
+    }
+
+    long retryDelayMs = 0;
+    long delayMultiplier = Duration.ofSeconds(5).toMillis();
+
+    // add delay only after the first failure
+    if (failCount > 1) {
+      retryDelayMs = delayMultiplier * Math.round(Math.pow(2, failCount - 2));
+    }
+    retryDelayMs = Math.min(retryDelayMs, clusterManagerConfig.getContainerRetryMaxDelayMs());
```

Comment:
Perhaps consider using org.apache.samza.util.ExponentialSleepStrategy or 
org.apache.commons.math3.distribution. ExponentialDistribution
to avoid dealing with backoff-match logic here.

Reply:
Cool, thanks.